Small fixes in PHP UploadHandler

smirgol opened the pull request:
I've fixed two small issues in the PHP UploadHandler:

get_config_bytes()
    remove trailing character from val

Technically it works the way it was before, but it will throw a "PHP Notice:  A non well formed numeric value encountered" when doing the multiplications between $val and the integers within the cases, as $val does contain a character at the end.

gd_create_scaled_image()
    fix case fallthrough in transparency handling

This was not working as expected for GIF images, as it would also enter the second case condition, which is intended for PNG files only. Probably not a problem at all, but I thought I might correct that.

Thank you for this great library, btw! :)

smirgol commented:
On the second commit https://github.com/blueimp/jQuery-File-Upload/pull/3523/commits/a6d67151014706d284516a0e615a2b43bd717636 my IDE cleaned up some white-spaces and corrected some indents, that's why there are more changes shown than I actually wanted to do. Sorry about that.

I changed some return values, where void was being returned, added an exception handling where imagick might throw an exception, initialized 3 vars that might be used in an uninitialized state and moved $response from the constructor to the class body. Nothing big, just some details. :-)

blueimp commented on server/php/UploadHandler.php:
```diff
@@ -424,12 +424,13 @@ protected function validate($uploaded_file, $file, $error, $index) {
             $file->error = $this->get_error_message('max_number_of_files');
             return false;
         }
+        $img_height = 0;
```
This should not be necessary, since `$img_height` is always set if `$img_width` is set.

See the following two lines:
https://github.com/blueimp/jQuery-File-Upload/blob/v9.30.0/server/php/UploadHandler.php#L439
https://github.com/blueimp/jQuery-File-Upload/blob/v9.30.0/server/php/UploadHandler.php#L454

For clarity however you might wanna change
```php
if (!empty($img_width)) {
```
to
```php
if (!empty($img_width) && !empty($img_height)) {
```

smirgol replied:
Yeah that it is a much better idea. I've ran into an issue with dividing by zero on friday with $y set to 0, when it was unable to load the image due to a wrong path and then not being able to read the image size properties. I'll look to also add a check if it actually loaded the image correctly before trying to do something with it.

blueimp commented on server/php/UploadHandler.php:
```diff
@@ -43,9 +43,9 @@ class UploadHandler
     const IMAGETYPE_PNG = 3;
 
     protected $image_objects = array();
+    protected $response = array();
```
Makes sense. ☺️

blueimp commented on server/php/UploadHandler.php:
```diff
@@ -795,25 +796,26 @@ protected function gd_create_scaled_image($file_name, $version, $options) {
                 imagecolortransparent($new_img, imagecolorallocate($new_img, 0, 0, 0));
+                break;
             case 'png':
+                imagecolortransparent($new_img, imagecolorallocate($new_img, 0, 0, 0));
```
Good catch! 👍 
I'm surprised there were no issues with executing  `imagealphablending` and `imagesavealpha` for `GIF` images before your fix.

blueimp commented on server/php/UploadHandler.php:
```diff
@@ -827,7 +829,11 @@ protected function imagick_get_image_object($file_path, $no_cache = false) {
+            try {
+                $image->readImage($file_path);
+            } catch (ImagickException $e) {
+                return null;
```
We should add an `error_log($e->getMessage());` before returning.

blueimp commented on server/php/UploadHandler.php:
```diff
@@ -896,32 +903,33 @@ protected function imagick_create_scaled_image($file_name, $version, $options) {
-        $new_height = $max_height = $img_height = $image->getImageHeight(); 
-		  
+        $new_height = $max_height = $img_height = $image->getImageHeight();
+        $x = $y = 0;
```
This should not be necessary, as in all cases where `$x` or `$y` are used, they're initialized first.

smirgol replied:
You're correct, sorry. My IDE went wild here I guess, showing me an issue where there actually isn't one. I didn't catch that.

blueimp replied:
No problem, but I'd definitely prefer to only add code lines that are necessary.

I also think the newlines don't really add much here and the convention for this source code file is to keep code inside of functions without extra newlines.

blueimp commented on server/php/UploadHandler.php:
```diff
@@ -1264,20 +1272,23 @@ protected function download() {
             default:
-                return $this->header('HTTP/1.1 403 Forbidden');
+                $this->header('HTTP/1.1 403 Forbidden');
+                return;
```
Any reason why you broke this up into multiple lines?
The `header` function already returns `null`.

blueimp commented on server/php/UploadHandler.php:
```diff
@@ -1264,20 +1272,23 @@ protected function download() {
         if (!$this->is_valid_file_object($file_name)) {
-            return $this->header('HTTP/1.1 404 Not Found');
+            $this->header('HTTP/1.1 404 Not Found');
+            return;
```
See previous comment.

blueimp commented on server/php/UploadHandler.php:
```diff
@@ -1264,20 +1272,23 @@ protected function download() {
                     true
                 )
             );
+            return;
```
See previous comment.

blueimp commented on server/php/UploadHandler.php:
```diff
@@ -1320,16 +1331,16 @@ public function generate_response($content, $print_response = true) {
             if ($redirect && preg_match($this->options['redirect_allow_target'], $redirect)) {
                 $this->header('Location: '.sprintf($redirect, rawurlencode($json)));
-                return;
+                return null;
```
`return` should be the same as `return null`.
I would understand it in this context if you want to make it more explicit that the content returned is `null`.

blueimp commented on server/php/UploadHandler.php:
```diff
@@ -1355,7 +1366,8 @@ public function head() {
         if ($print_response && $this->get_query_param('download')) {
-            return $this->download();
+            $this->download();
+            return null;
```
The `download` function already returns `null`.

smirgol replied:
download() does not return anything, so it's a void, which cannot be returned.
Admittedly I was a bit unsure on what to return here instead, I figured a null would be the best, but I have to trace where the returned value of get() is being used, to see if it will be tested against being null.

smirgol commented:
Thank you very much for your feedback. I'll look into resolving these issues tomorrow.

About void and null: they are not the same. void is really nothing, null is actually a value - a null value.
A "null" can be returned, a "void" cannot (at least in C/CPP, PHP still does allow it, as void implementation is incomplete).

The header() function for example does not return anything (void), so you can/should not return it's not existing return value like "return header('some header')". That does work in PHP, but is actually not valid, since you can return only something if you have something. That's why I split these calls up and added a "return" in the follow up line - which returns from the method and does not return anything (void).

Hope that makes sense. :)

smirgol commented on server/php/UploadHandler.php:
```diff
@@ -377,7 +377,7 @@ protected function get_error_message($error) {
         $val = trim($val);
         $last = strtolower($val[strlen($val)-1]);
-        $val = (int)$val;
+        $val = (int)substr($val, 0, -1);
```
Good catch, I've actually added that already, but missed to push the commit.

blueimp commented on server/php/UploadHandler.php:
```diff
@@ -377,7 +377,11 @@ protected function get_error_message($error) {
         $val = trim($val);
         $last = strtolower($val[strlen($val)-1]);
-        $val = (int)substr($val, 0, -1);
+        if (false == preg_match("/[0-9]$/", $val)) {
```
I think this can be improved by using the built-in `is_numeric` function:
```php
if (is_numeric($last)) {
    $val = (int)$val;
} else {
    $val = (int)substr($val, 0, -1);
} 
```

smirgol replied:
True, I did have some problems with is_numeric in the past, but that was a long time ago, admittedly. It seems to do it's job nowadays. :)

blueimp commented on server/php/UploadHandler.php:
```diff
@@ -452,7 +455,7 @@ function_exists('exif_read_data') &&
             }
         }
-        if (!empty($img_width)) {
+        if (!empty($img_width) && !empty($img_height)) {
```
Yeah I think this is a good improvement as it adds clarity. 👍

blueimp commented on server/php/UploadHandler.php:
```diff
@@ -832,6 +835,7 @@ protected function imagick_get_image_object($file_path, $no_cache = false) {
             try {
                 $image->readImage($file_path);
             } catch (ImagickException $e) {
+                error_log($e->getMessage());
```
Thanks for logging the error. ☺️

blueimp commented:
Thanks again @smirgol! ☺️
